svm/src/transaction_processor.rs

```diff
@@ -428,29 +428,32 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
                          lamports_per_signature,
                      }| {
                         let message = sanitized_tx.borrow().message();
-                        let (fee_details, fee_payer_account, fee_payer_rent_debit) = self
-                            .validate_transaction_fee_payer(
-                                callbacks,
-                                message,
-                                feature_set,
-                                lamports_per_signature,
-                                rent_collector,
-                                error_counters,
-                            )?;
-
-                        // Update nonce with fee-subtracted accounts
-                        let fee_payer_address = message.fee_payer();
-                        let nonce = nonce.map(|nonce| {
-                            NonceFull::from_partial(
-                                nonce,
-                                fee_payer_address,
-                                fee_payer_account.clone(),
-                                fee_payer_rent_debit,
-                            )
-                        });
+                        let (
+                            fee_details,
+                            fee_payer_account,
+                            fee_payer_rent_debit,
+                            fee_payer_rent_epoch,
+                        ) = self.validate_transaction_fee_payer(
+                            callbacks,
+                            message,
+                            feature_set,
+                            lamports_per_signature,
+                            rent_collector,
+                            error_counters,
+                        )?;
+
+                        // Capture fee-subtracted fee payer account and original nonce account state
+                        // to rollback to if transaction execution fails.
+                        let rollback_accounts = RollbackAccounts::new(
```

Comment:
I think `RollbackAccounts::new()` should be moved inside `::validate_transaction_fee_payer()` by newly passing `nonce` to it.

In this way, the special casing of `fee_payer_account.set_lamports` and `fee_payer_account.set_rent_epoch()` can be erased inside `RollbackAccounts::new()`, along with `fee_payer_rent_debit` and `fee_payer_loaded_rent_epoch` args removed.

Reply:
the underlying motive is that it's a bit awkward conceptually, for `RollbackAccounts` to fiddles states just to restore the _real_ rollback-able accounts. Instead, it should be constructed at the right place where the rollback-able account exists.

Reply:
It doesn't actually change much to move `RollbackAccounts::new` inside `validate_transaction_fee_payer`. We still need to deduct rent before fees to match current runtime behavior. And as long as we collect rent first, we need to have some logic to credit rent and reset the rent epoch. I think it's nice to have this fiddling inside the `RollbackAccounts::new` function so that it's self contained for now. We can clean this up later (probably with some feature gates) when rent is disabled on all of the clusters.

Reply:
FYI I moved `RollbackAccounts::new` into `validate_transaction_fee_payer` because I think it does make sense for it to live there and it simplifies `validate_fees` a lot.

Reply:
> Instead, it should be constructed at the right place where the rollback-able account exists.

Unfortunately due to rent, there's never really a point where a rollback-able account exists unless we explicitly create one. This is because the process to load the fee payer account is:
1. Collect rent
2. Collect fees

And the process to create a rollback account is
1. Collect fees
2. Update rent epoch if it's a nonce transaction

Since loading the fee payer account requires that we do rent collection first, there's never a point during normal loading that the rollback account is created. Not to mention that we have to update the rent epoch only for certain types of transactions.

We'll have to wait until rent collection / rent epoch modifications are cleaned up before making this implementation any cleaner.